text/src/tail.rs

```diff
@@ -149,53 +177,57 @@ fn tail(args: &Args) -> Result<(), Box<dyn std::error::Error>> {
     // If follow option is specified, continue monitoring the file
     if args.follow && !(args.file == Some(PathBuf::from("-")) || args.file.is_none()) {
         let file_path = args.file.as_ref().unwrap();
-        // Initialization of inotify
-        let mut inotify = Inotify::init()?;
-
-        inotify
-            .watches()
-            .add(file_path, WatchMask::MODIFY | WatchMask::DELETE_SELF)
-            .expect("Failed to add inotify watch");
-
-        // Opening a file and placing the cursor at the end of the file
-        let mut file = File::open(file_path)?;
-        file.seek(SeekFrom::End(0))?;
-        let mut reader = BufReader::new(file);
-
-        // Buffer for inotify events
-        let mut buffer = [0u8; 4096];
-
-        loop {
-            // Read inotify events
-            let events = inotify.read_events_blocking(&mut buffer)?;
-
-            // Handle each event
-            for event in events {
-                if event.mask.contains(inotify::EventMask::DELETE_SELF) {
-                    eprintln!("File {} deleted, exiting...", file_path.display());
-                    std::process::exit(1);
-                }
-
-                if event.mask.contains(inotify::EventMask::MODIFY) {
-                    // If the file has been modified, check if the file was truncated
-                    let metadata = fs::metadata(file_path)?;
-                    let current_size = metadata.len();
 
-                    if current_size < reader.stream_position()? {
-                        eprintln!("tail: {}: file truncated", file_path.display());
-
-                        reader.seek(SeekFrom::Start(0))?;
-                        reader = BufReader::new(File::open(file_path)?);
-                    }
-
-                    // Read the new lines and output them
-                    let mut new_data = String::new();
-                    let bytes_read = reader.read_to_string(&mut new_data)?;
-                    if bytes_read > 0 {
-                        print!("{}", new_data);
-                        io::stdout().flush()?;
+        let mut last_known_contents = read_file_to_bytes(file_path)?;
+        let mut last_known_hash = calculate_hash(&last_known_contents);
```

Comment:
No need in hashing or storing the whole file in memory. This would fill the whole memory with large files and crash